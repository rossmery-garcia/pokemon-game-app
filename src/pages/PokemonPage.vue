<template>
  <h2 v-if="!pokemon">Espere por favor...</h2>

  <div v-else>
    <h1>¿Quién es este Pokémon?</h1>

    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"
    />

    <PokemonOptions
      :pokemons="pokemonArr"
      @selectionPokemon="checkAnswer"
    />

    <template v-if="showAnswer" >
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>

</template>

<script>
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';

import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    }
  },
  methods: {
    async mixPokemonArr() {
      this.pokemonArr = await getPokemonOptions();

      const randomInt = Math.floor( Math.random() * 4 );
      this.pokemon = this.pokemonArr[ randomInt ];
    },
    checkAnswer( selectedId ) {
      this.showPokemon = true;
      this.showAnswer = true;

      if( selectedId === this.pokemon.id ) {
        this.message = `Correcto, ${ this.pokemon.name }`;
      } else {
        this.message = `Oops, era ${ this.pokemon.name }`;
      }
    },
    newGame() {
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];

      this.mixPokemonArr();
    }
  },
  mounted() {
    this.mixPokemonArr();
  }
}
</script>

<style>

</style>