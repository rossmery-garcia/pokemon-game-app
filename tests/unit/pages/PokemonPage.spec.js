import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { mockPokemons } from '../mocks/pokemons.mock';

describe('PokemonPage component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount( PokemonPage );
  });

  test('Should match with the snapshot', () => {
    expect( wrapper.html() ).toMatchSnapshot();
  });

  test('Should call mixPokemonArr on mount', () => {

    const mixPokemonArrSpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArr');
    wrapper = shallowMount( PokemonPage );

    expect( mixPokemonArrSpy ).toHaveBeenCalled();
  });

  test('Should match the snapshot when loading pokemon', () => {

    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        }
      }
    });

    expect( wrapper.html() ).toMatchSnapshot();
  });

  test('Should show PokemonPicture and PokemonOptions components', () => {

    const wrapper = shallowMount( PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        }
      }
    });

    const pokemonPictureComponent = wrapper.find('pokemon-picture-stub');
    const pokemonOptionsComponent = wrapper.find('pokemon-options-stub');

    expect( pokemonPictureComponent.exists() ).toBeTruthy();
    expect( pokemonOptionsComponent.exists() ).toBeTruthy();

    expect( pokemonPictureComponent.attributes('pokemonid') ).toBe('1');
    expect( pokemonOptionsComponent.attributes('pokemons') ).toBeTruthy();
  });
})