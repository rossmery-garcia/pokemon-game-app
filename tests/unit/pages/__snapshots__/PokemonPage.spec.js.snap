// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Should match the snapshot when loading pokemon 1`] = `
<div>
  <h1>¿Quién es este Pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage component Should match with the snapshot 1`] = `<h2>Espere por favor...</h2>`;
